iree/compiler/Conversion/HLOToLinalg/FusionOfTensorOps.cpp

```diff
@@ -89,7 +89,22 @@ struct FusionOfTensorOpsPass
     // indexed generic operations. If an operation is used in a named op, it
     // will be computed anyway, so the consumers can just use that value.
     linalg::ControlElementwiseOpsFusionFn controlFn =
-        [](const OpResult &producer, const OpOperand & /*consumer*/) {
+        [](const OpResult &producer, const OpOperand &consumer) {
+          // TODO(ataei): Enable fusion with reduction consumer. Currently
+          // vectorization doesn't handle generic ops with reduction iterators
+          // we will disable for now to allow vectorizing producer pointwise
+          // ops.
+          auto consumerOp = consumer.getOwner();
+          if (isa<linalg::GenericOp, linalg::IndexedGenericOp>(consumerOp)) {
+            auto linalgOp = dyn_cast<linalg::LinalgOp>(consumerOp);
```

Comment:
You could just make this a `<cast<linalg::LinalgOp>`

Reply:
This will disable fusion with matmul 

> Sorry not trying to be annoying and blocking on fixing CPU regression issues. :D But with this we are creating more dispatch regions, where on GPU they actually can be fused. What about adding a command-line option so I can at least add it to benchmarking configurations. I don't care the default value to be on or off; you can feel free to choose. :)

Sure I will expose this to be optionally with a flag. But my intuition this should be faster on the GPU side too. Invoking more dispatches with producer that is parallel point wise should be faster than fusing the consumer into that producer and run the reduction serially. see this example:

```
func @fused_region(%4: tensor<4x384x384xf32>, %8: tensor<384xi32>, %6: tensor<4x384xf32>, %7: tensor<4x384xf32>) -> tensor<4x384xf32> attributes { iree.module.export } {
    %cst = constant 0.176776692 : f32
    %cst_0 = constant -1.000000e+04 : f32
    %cst_1 = constant 1.000000e+04 : f32
    %cst_2 = constant 1.000000e+00 : f32
    %9 = linalg.generic {
        indexing_maps = [
          affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d2)>, 
          affine_map<(d0, d1, d2) -> (d0, d1)>, affine_map<(d0, d1, d2) -> (d0, d1)>
        ], 
        iterator_types = ["parallel", "parallel", "reduction"]
      } ins(%4, %8, %6 : tensor<4x384x384xf32>, tensor<384xi32>, tensor<4x384xf32>) outs(%7 : tensor<4x384xf32>) {
      ^bb0(%arg0: f32, %arg1: i32, %arg2: f32, %arg3: f32):  // no predecessors
        %10 = sitofp %arg1 : i32 to f32
        %11 = mulf %10, %cst_2 : f32
        %12 = mulf %11, %cst_1 : f32
        %13 = addf %12, %cst_0 : f32
        %14 = mulf %arg0, %cst : f32
        %15 = addf %14, %13 : f32
        %16 = subf %15, %arg2 : f32
        %17 = math.exp %16 : f32
        %18 = addf %17, %arg3 : f32
        linalg.yield %18 : f32
      } -> tensor<4x384xf32>
    return %9 : tensor<4x384xf32>
}
```

Reply:
here is the reduction part
```
 %18 = addf %17, %arg3 : f32
 ```
 with this change this is going to be two dispatches

Reply:
Maybe we should hold off on adding the flag. If it causes "significant" regression on GPU, then it makes sense. @antiagainst two options here
1) Let this land and we can see the impact on performance. Add the flag later.
2) If you have data to show this regresses (significantly) we can add flag here.

I really want to avoid having these flags. Those will become unsustainable.

Reply:
It's more nuanced I'm afraid. There are actually another aspect to this: GPU in general has much more parallelism than CPU. Two dispatches means a barrier in the middle and a barrier typically means draining the whole GPU pipeline and then refilling it. The cost to ramp down and then ramp up again can be big, esp. for element wise ops where both kernels won't quite a long time to complete. :)

Reply:
This is running all the math above serially. Its very surprising to me that cost of 4x384x384/(GPU_VECTOR_SIZE) ~10 add/mul + exp ops < cost issuing a dispatch!!

Reply:
Didn't see @MaheshRavishankar's comments before the previous reply. I suggested to have a flag given this is a temporary fix (there is a TODO for it) and I'd assume it will be fixed and remove later. I will try to see the impacts on the phone but I actually don't expect "significant" regression. Don't want to be blocking here; so please feel free to land as-is. :)

Reply:
In any case we really need to enable running mako or (something else) on PRs. 
cc @hanhanW @GMNGeoffrey for how can we do that

Reply:
Mako is just a place to upload benchmark results. We deliberately don't upload results from PRs to the main benchmark because that would pollute the data. I believe there is such a thing as a "dev" benchmark or something that has lower restrictions (or maybe that's just a concept we can implement). @hanhanW could you perhaps set up a separate benchmark for each thing we're measuring that doesn't have any regression analysis, but allows uploading results? This would allow testing the full pipeline (including upload) and viewing the results on Mako. In the meantime @asaadaldien you can run the benchmark pipeline on a PR and look at the benchmark results in the log.

Unless you're talking about *automatic* regression monitoring for a PR (like just a status check that reports a regression)? That is a larger task. Currently we don't even run Android correctness tests on presubmit because they're using real Android phones which are flakey and constrained. I proposed that we should be also benchmarking on cloud VMs to catch general regressions in a space we can easily scale and replicate. Not sure whether Mako's regression testing would work with that (I'm overall not super impressed with it tbh).

Reply:
@asaadaldien: IIUC of the example you have here, the generic output uses two parallel dimensions so we should be able to distribute along these two dimensions? I don't think it's fully serialized? For each distributed thread yes it's serialized because we have reduction dimension inside. Also the above is just one example; the change applies broader right? :)

Anyway, again, not trying to be annoying here. And I'm sorry if I sounded negative. What I'm actually wanted to see is this kind of things can be nuanced. It needs proper measuring to see.

Please feel free to merge.

Reply:
I think its more about running the benchmarks that are tracked on Mako on a PR to catch regression before submitting the PR. Most of us have android phones, but not ones with a Mali G77. So we can verify the performance locally on CPU, but not GPUs. I think just triggering a run on the build bot (here : https://buildkite.com/iree/iree-android-benchmark/builds/) and seeing the results should be enough for this.

Reply:
Kicked off a build here https://buildkite.com/iree/iree-android-benchmark/builds/2302

Reply:
Cool! that is very close to what I was looking for.

Reply:
Regarding benchmarking: right now we only have one S20 and one pixel 4 in the lab so it will be the bottleneck. But it's more due to last year's lack of budget. I think we have more budget for this year now. We should by more phones and attach them to the lab and run even on PRs. Let me chat with Julian about this later.